Extension/lib/utils/element-collapser.js

```diff
@@ -52,37 +54,83 @@ var ElementCollapser = (function() {
+     * @param rule css rule
+     * @param shadowRoot
      */
-    var hideBySelector = function(selectorText, cssText) {
-        var styleElement = document.getElementById(collapserStyleId);
+    var applyCss = function (rule, shadowRoot) {
+        var styleElement = getStyleElement(shadowRoot);
         if (!styleElement) {
             styleElement = document.createElement("style");
             styleElement.id = collapserStyleId;
             styleElement.setAttribute("type", "text/css");
-            (document.head || document.documentElement).appendChild(styleElement);
+
+            if (shadowRoot) {
+                shadowRoot.appendChild(styleElement);
+            } else {
+                (document.head || document.documentElement).appendChild(styleElement);
+            }
+        }
+
+        if (shadowRoot) {
+            rule = '::content ' + rule;
         }
 
-        var rule = selectorText + '{' + (cssText || "display: none!important") + '}';
         styleElement.sheet.insertRule(rule, styleElement.sheet.cssRules.length);
     };
 
     /**
+     * Adds "selectorText { display:none!important; }" style
+     *
+     * @param selectorText
+     * @param cssText optional
+     * @param shadowRoot optional
+     */
+    var hideBySelector = function(selectorText, cssText, shadowRoot) {
+        var rule = selectorText + '{' + (cssText || "display: none!important;") + '}';
+
+        applyCss(rule, shadowRoot);
+    };
+
+    /**
+     * Adds "selectorText { display:none!important; }" style
+     *
+     * @param shadowRoot optional
+     */
+    var hideBySelectorAndTagName = function(selectorText, tagName, shadowRoot) {
+        if (tagName === "frame" || tagName === "iframe") {
+            // Use specific style for frames due to these issues:
+            // https://github.com/AdguardTeam/AdguardBrowserExtension/issues/346
+            // https://github.com/AdguardTeam/AdguardBrowserExtension/issues/355
+            // https://github.com/AdguardTeam/AdguardBrowserExtension/issues/347
+            hideBySelector(selectorText, "visibility: hidden!important; height: 0px!important;", shadowRoot);
+        } else {
+            hideBySelector(selectorText, null, shadowRoot);
+        }
+    };
+
+    /**
      * Unhides elements which were previously hidden by the specified selector
      */
-    var unhideBySelector = function(selectorText) {
-        var styleElement = document.getElementById(collapserStyleId);
+    var unhideBySelector = function(selectorText, shadowRoot) {
+        var styleElement = getStyleElement(shadowRoot);
         if (!styleElement || !styleElement.sheet) {
             return;
         }
         var iLength = styleElement.sheet.cssRules.length;
         while (iLength--) {
             var cssRule = styleElement.sheet.cssRules[iLength];
-            if (cssRule.selectorText == selectorText) {
+            if (cssRule.selectorText == (shadowRoot ? '::content ' + selectorText : selectorText)) {
```

Comment:
Can be moved to a separate helper function `prepareSelectorText(selectorText, shadowRoot)`